src/common/apple/purchase.ts

```diff
@@ -0,0 +1,247 @@
+};
+
+export const handleCoresPurchase = async ({
+  transactionInfo,
+  user,
+}: {
+  transactionInfo: JWSTransactionDecodedPayload;
+  user: Pick<User, 'id' | 'subscriptionFlags' | 'coresRole'>;
+  environment: Environment;
+  notification: ResponseBodyV2DecodedPayload;
+}): Promise<UserTransaction> => {
+  if (!transactionInfo.transactionId) {
+    throw new Error('Missing transactionId in transactionInfo');
+  }
+
+  if (!transactionInfo.productId) {
+    throw new Error('Missing productId in transactionInfo');
+  }
+
+  if (
+    checkUserCoresAccess({
+      user,
+      requiredRole: CoresRole.User,
+    }) === false
+  ) {
+    throw new Error('User does not have access to cores purchase');
+  }
+
+  const con = await createOrGetConnection();
+
+  // TODO feat/cores-iap load from api metadata/new endpoint
+  const coresValue = Number(transactionInfo.productId.match(/\d+/)?.[0]);
+
+  const payload = con.getRepository(UserTransaction).create({
+    processor: UserTransactionProcessor.AppleStoreKit,
+    receiverId: user.id,
+    status: UserTransactionStatus.Success,
+    productId: null, // no product user is buying cores directly
+    senderId: null, // no sender, user is buying cores
+    value: coresValue,
+    valueIncFees: coresValue,
+    fee: 0, // no fee when buying cores
+    request: {},
+    flags: {
+      providerId: transactionInfo.transactionId,
+    },
+  });
+
+  const transaction = await con.transaction(async (entityManager) => {
+    const userTransaction = await entityManager
+      .getRepository(UserTransaction)
+      .save(payload);
+
+    // TODO feat/cores-iap enable for production https://dailydotdev.slack.com/archives/C07VA1FJTDK/p1745580651217029
+    if (!isProd) {
+      try {
+        await purchaseCores({
+          transaction: userTransaction,
+        });
+      } catch (error) {
+        if (error instanceof TransferError) {
+          const userTransactionError = new UserTransactionError({
+            status: error.transfer.status,
+            transaction: userTransaction,
+          });
+
+          // update transaction status to error
+          await entityManager.getRepository(UserTransaction).update(
+            {
+              id: userTransaction.id,
+            },
+            {
+              status: error.transfer.status as number,
+              flags: updateFlagsStatement<UserTransaction>({
+                error: userTransactionError.message,
+              }),
+            },
+          );
+
+          return entityManager.getRepository(UserTransaction).create({
+            ...userTransaction,
+            status: error.transfer.status as number,
+            flags: {
+              ...userTransaction.flags,
+              error: userTransactionError.message,
+            },
+          });
+        }
+
+        throw error;
+      }
+    }
+
+    return userTransaction;
+  });
+
+  if (transaction.status === UserTransactionStatus.Success) {
```

Comment:
I think we should also log an event here just like we do with paddle, or does it happen on the ios part?

Reply:
yeah, missed calling the function, adding

Reply:
https://github.com/dailydotdev/daily-api/pull/2783/commits/c57a4b3f5e14b56e922f70c271d180b3d82ab893 added